File: tests/e2e/go_client/g2_upload_endpoints_test.go
// G2 upload-host endpoints E2E against a live bqemulator container.
//
// The Go ``bigquery.NewClient(...).Dataset(...).Table(...).LoaderFrom(reader)``
// API goes through the upload-host endpoints — this test exercises
// both multipart (small payload) and resumable (large payload) paths.

package e2e

import (
	"bytes"
	"context"
	"fmt"
	"strings"
	"testing"
	"time"

	"cloud.google.com/go/bigquery"
	"google.golang.org/api/iterator"
	"google.golang.org/api/option"
)

const (
	g2Project = "e2e-go-g2"
)

func g2Client(ctx context.Context, t *testing.T) *bigquery.Client {
	t.Helper()
	client, err := bigquery.NewClient(
		ctx,
		g2Project,
		option.WithEndpoint(bqAPIBase()),
		option.WithoutAuthentication(),
	)
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	return client
}

func TestG2LoadCSVMultipart(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	client := g2Client(ctx, t)
	defer client.Close()

	ds := client.Dataset("g2_go_csv")
	_ = ds.DeleteWithContents(ctx)
	if err := ds.Create(ctx, &bigquery.DatasetMetadata{Location: "US"}); err != nil {
		t.Fatalf("create dataset: %v", err)
	}
	defer func() { _ = ds.DeleteWithContents(ctx) }()

	tbl := ds.Table("rows")
	if err := tbl.Create(ctx, &bigquery.TableMetadata{
		Schema: bigquery.Schema{
			{Name: "id", Type: bigquery.IntegerFieldType},
			{Name: "name", Type: bigquery.StringFieldType},
		},
	}); err != nil {
		t.Fatalf("create table: %v", err)
	}

	csv := []byte("id,name\n1,alice\n2,bob\n3,carol\n4,dan\n")
	rs := bigquery.NewReaderSource(bytes.NewReader(csv))
	rs.SourceFormat = bigquery.CSV
	rs.SkipLeadingRows = 1
	rs.Schema = bigquery.Schema{
		{Name: "id", Type: bigquery.IntegerFieldType},
		{Name: "name", Type: bigquery.StringFieldType},
	}
	loader := tbl.LoaderFrom(rs)
	loader.WriteDisposition = bigquery.WriteTruncate

	job, err := loader.Run(ctx)
	if err != nil {
		t.Fatalf("loader run: %v", err)
	}
	status, err := job.Wait(ctx)
	if err != nil {
		t.Fatalf("loader wait: %v", err)
	}
	if status.Err() != nil {
		t.Fatalf("loader err: %v", status.Err())
	}

	q := client.Query(fmt.Sprintf("SELECT COUNT(*) AS n FROM `%s.g2_go_csv.rows`", g2Project))
	it, err := q.Read(ctx)
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	var row struct{ N int64 }
	if err := it.Next(&row); err != nil && err != iterator.Done {
		t.Fatalf("iterate: %v", err)
	}
	if row.N != 4 {
		t.Fatalf("want 4 rows, got %d", row.N)
	}
}

func TestG2LoadNDJSONResumable(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 120*time.Second)
	defer cancel()
	client := g2Client(ctx, t)
	defer client.Close()

	ds := client.Dataset("g2_go_json")
	_ = ds.DeleteWithContents(ctx)
	if err := ds.Create(ctx, &bigquery.DatasetMetadata{Location: "US"}); err != nil {
		t.Fatalf("create dataset: %v", err)
	}
	defer func() { _ = ds.DeleteWithContents(ctx) }()

	tbl := ds.Table("rows")
	if err := tbl.Create(ctx, &bigquery.TableMetadata{
		Schema: bigquery.Schema{
			{Name: "id", Type: bigquery.IntegerFieldType},
			{Name: "name", Type: bigquery.StringFieldType},
		},
	}); err != nil {
		t.Fatalf("create table: %v", err)
	}

	// Build a ~2 MiB NDJSON payload to push the client into the
	// resumable upload protocol.
	var b strings.Builder
	for i := 0; i < 60000; i += 1 {
		fmt.Fprintf(&b, "{\"id\":%d,\"name\":\"name-%d\"}\n", i, i)
	}
	ndjson := b.String()
	if len(ndjson) < 1_000_000 {
		t.Fatalf("payload too small: %d", len(ndjson))
	}

	rs := bigquery.NewReaderSource(strings.NewReader(ndjson))
	rs.SourceFormat = bigquery.JSON
	rs.Schema = bigquery.Schema{
		{Name: "id", Type: bigquery.IntegerFieldType},
		{Name: "name", Type: bigquery.StringFieldType},
	}
	loader := tbl.LoaderFrom(rs)
	loader.WriteDisposition = bigquery.WriteTruncate

	job, err := loader.Run(ctx)
	if err != nil {
		t.Fatalf("loader run: %v", err)
	}
	status, err := job.Wait(ctx)
	if err != nil {
		t.Fatalf("loader wait: %v", err)
	}
	if status.Err() != nil {
		t.Fatalf("loader err: %v", status.Err())
	}

	q := client.Query(fmt.Sprintf("SELECT COUNT(*) AS n FROM `%s.g2_go_json.rows`", g2Project))
	it, err := q.Read(ctx)
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	var row struct{ N int64 }
	if err := it.Next(&row); err != nil && err != iterator.Done {
		t.Fatalf("iterate: %v", err)
	}
	if row.N != 60000 {
		t.Fatalf("want 60000 rows, got %d", row.N)
	}
}

func TestLoadTableCSVAutodetect(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	client := g2Client(ctx, t)
	defer client.Close()

	dsID := "g2_csv_autodetect"
	tblID := "rows"

	ds := client.Dataset(dsID)
	_ = ds.DeleteWithContents(ctx)
	if err := ds.Create(ctx, &bigquery.DatasetMetadata{Location: "US"}); err != nil {
		t.Fatalf("Failed to create dataset: %v", err)
	}
	defer func() { _ = ds.DeleteWithContents(ctx) }()

	csvData := []byte("id,name,score\n1,alice,99.5\n2,bob,88.2\n")
	rs := bigquery.NewReaderSource(bytes.NewReader(csvData))
	rs.SourceFormat = bigquery.CSV
	rs.SkipLeadingRows = 1
	rs.AutoDetect = true
	
	loader := ds.Table(tblID).LoaderFrom(rs)
	loader.WriteDisposition = bigquery.WriteTruncate
	loader.CreateDisposition = bigquery.CreateIfNeeded

	job, err := loader.Run(ctx)
	if err != nil {
		t.Fatalf("Loader.Run failed: %v", err)
	}
	status, err := job.Wait(ctx)
	if err != nil {
		t.Fatalf("Job wait failed: %v", err)
	}
	if err := status.Err(); err != nil {
		t.Fatalf("Job failed: %v", err)
	}

	q := client.Query(fmt.Sprintf("SELECT COUNT(*) AS n FROM `%s.%s.%s`", g2Project, dsID, tblID))
	it, err := q.Read(ctx)
	if err != nil {
		t.Fatalf("Query failed: %v", err)
	}

	var row struct{ N int64 }
	err = it.Next(&row)
	if err != iterator.Done && err != nil {
		t.Fatalf("Next failed: %v", err)
	}
	if row.N != 2 {
		t.Errorf("Expected 2 rows, got %d", row.N)
	}

	// Verify inferred schema via Metadata
	meta, err := ds.Table(tblID).Metadata(ctx)
	if err != nil {
		t.Fatalf("Failed to fetch metadata: %v", err)
	}
	if len(meta.Schema) != 3 {
		t.Fatalf("Expected 3 schema fields, got %d", len(meta.Schema))
	}
	if meta.Schema[0].Name != "id" || (meta.Schema[0].Type != bigquery.IntegerFieldType && meta.Schema[0].Type != "INT64") {
		t.Errorf("Expected id INT64, got %v %v", meta.Schema[0].Name, meta.Schema[0].Type)
	}
	if meta.Schema[1].Name != "name" || (meta.Schema[1].Type != bigquery.StringFieldType && meta.Schema[1].Type != "STRING") {
		t.Errorf("Expected name STRING, got %v %v", meta.Schema[1].Name, meta.Schema[1].Type)
	}
	if meta.Schema[2].Name != "score" || (meta.Schema[2].Type != bigquery.FloatFieldType && meta.Schema[2].Type != "FLOAT64") {
		t.Errorf("Expected score FLOAT64, got %v %v", meta.Schema[2].Name, meta.Schema[2].Type)
	}
}

File: tests/e2e/nodejs_client/test/g2_upload_endpoints.test.js
/**
 * E2E: G2 upload-host endpoints (multipart + resumable) against a live
 * bqemulator container via the @google-cloud/bigquery Node.js client.
 *
 * The Node client's `table.load(streamOrFile)` API auto-selects
 * multipart vs resumable upload based on the payload size. Both
 * branches exercise `/upload/bigquery/v2/projects/{p}/jobs` rather
 * than the data-plane `/jobs` endpoint.
 */

const { describe, it, before, after } = require("node:test");
const assert = require("node:assert/strict");
const { Readable } = require("node:stream");

const REST_URL = process.env.BQEMU_REST_URL || "http://localhost:9050";
const PROJECT = "e2e-nodejs-g2";
const DATASET = "g2_node_ds";

function makeClient() {
  const { BigQuery } = require("@google-cloud/bigquery");
  const { OAuth2Client } = require("google-auth-library");

  const fake = new OAuth2Client();
  fake.credentials = { access_token: "anonymous" };

  return new BigQuery({
    projectId: PROJECT,
    apiEndpoint: REST_URL,
    authClient: fake,
    autoRetry: false,
  });
}

async function cleanup(client) {
  try {
    await client.dataset(DATASET).delete({ force: true });
  } catch (_) {
    /* ignore */
  }
}

describe("bqemulator G2 upload host (Node.js)", () => {
  let client;
  before(async () => {
    client = makeClient();
    await client.createDataset(DATASET, { location: "US" }).catch(() => {});
  });
  after(async () => {
    await cleanup(client);
  });

  it("loads a small CSV via multipart upload", async () => {
    await client.query(
      `CREATE TABLE IF NOT EXISTS \`${PROJECT}.${DATASET}.rows_mp\` ` +
        `(id INT64, name STRING)`,
    );

    const csv = Buffer.from("id,name\n1,alice\n2,bob\n3,carol\n", "utf-8");

    // @google-cloud/bigquery@7 requires a File object for table.load(),
    // not a Buffer / Readable. Use createWriteStream() to pipe arbitrary
    // bytes into the upload host — this exercises the same multipart
    // endpoint without depending on @google-cloud/storage.
    const writeStream = client
      .dataset(DATASET)
      .table("rows_mp")
      .createWriteStream({
        sourceFormat: "CSV",
        skipLeadingRows: 1,
        writeDisposition: "WRITE_TRUNCATE",
        schema: {
          fields: [
            { name: "id", type: "INTEGER" },
            { name: "name", type: "STRING" },
          ],
        },
      });
    // The "job" event fires with the Job once the upload is accepted;
    // "complete" fires when the upload stream is fully written. After
    // both fire, poll the Job's metadata for terminal state — the
    // upload-host endpoint synchronously transitions to DONE on the
    // emulator, so a single getMetadata() round-trip is enough.
    const job = await new Promise((resolve, reject) => {
      let receivedJob = null;
      writeStream.on("job", (j) => {
        receivedJob = j;
      });
      writeStream.on("complete", () => resolve(receivedJob));
      writeStream.on("error", reject);
      Readable.from(csv).pipe(writeStream);
    });
    const [meta] = await job.getMetadata();
    assert.equal(meta.status.state, "DONE");

    const [rows] = await client.query(
      `SELECT COUNT(*) AS n FROM \`${PROJECT}.${DATASET}.rows_mp\``,
    );
    assert.equal(Number(rows[0].n), 3);
  });

  it("loads a larger NDJSON via resumable upload", async () => {
    await client.query(
      `CREATE TABLE IF NOT EXISTS \`${PROJECT}.${DATASET}.rows_rs\` ` +
        `(id INT64, name STRING)`,
    );

    // Synthesize ~2 MiB of NDJSON to push the client into resumable mode.
    const lines = [];
    for (let i = 0; i < 60_000; i += 1) {
      lines.push(JSON.stringify({ id: i, name: `name-${i}` }));
    }
    const ndjson = Buffer.from(lines.join("\n") + "\n", "utf-8");
    assert.ok(ndjson.length > 1_000_000);

    // See the multipart test above for why we use createWriteStream
    // instead of table.load(stream, ...) — v7 of the BigQuery client
    // requires a @google-cloud/storage File object for direct load().
    const writeStream = client
      .dataset(DATASET)
      .table("rows_rs")
      .createWriteStream({
        sourceFormat: "NEWLINE_DELIMITED_JSON",
        writeDisposition: "WRITE_TRUNCATE",
        schema: {
          fields: [
            { name: "id", type: "INTEGER" },
            { name: "name", type: "STRING" },
          ],
        },
      });
    const job = await new Promise((resolve, reject) => {
      let receivedJob = null;
      writeStream.on("job", (j) => {
        receivedJob = j;
      });
      writeStream.on("complete", () => resolve(receivedJob));
      writeStream.on("error", reject);
      Readable.from(ndjson).pipe(writeStream);
    });
    const [meta] = await job.getMetadata();
    assert.equal(meta.status.state, "DONE");

    const [rows] = await client.query(
      `SELECT COUNT(*) AS n FROM \`${PROJECT}.${DATASET}.rows_rs\``,
    );
    assert.equal(Number(rows[0].n), 60_000);
  });
  it("should honor autodetect flag for CSV load", async () => {
    const tableId = "rows_autodetect";
    const csvData = Buffer.from("id,name,score\n1,alice,99.5\n2,bob,88.2\n", "utf-8");

    const writeStream = client
      .dataset(DATASET)
      .table(tableId)
      .createWriteStream({
        sourceFormat: "CSV",
        skipLeadingRows: 1,
        autodetect: true,
        writeDisposition: "WRITE_TRUNCATE",
        createDisposition: "CREATE_IF_NEEDED",
      });

    const job = await new Promise((resolve, reject) => {
      let receivedJob = null;
      writeStream.on("job", (j) => {
        receivedJob = j;
      });
      writeStream.on("complete", () => resolve(receivedJob));
      writeStream.on("error", reject);
      Readable.from(csvData).pipe(writeStream);
    });

    const [meta] = await job.getMetadata();
    assert.equal(meta.status.state, "DONE");

    const [rows] = await client.query(
      `SELECT COUNT(*) AS n FROM \`${PROJECT}.${DATASET}.${tableId}\``
    );
    assert.equal(Number(rows[0].n), 2);
  });
});
